File: source/cpp/145. Binary Tree Postorder Traversal.cpp
#include <stack>
#include <vector>


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    std::vector<int> postorderTraversal(const TreeNode* root) const {
        if (root == nullptr) return {};

        std::vector<int> result;
        std::stack<const TreeNode*> stack;
        stack.push(root);

        while (!stack.empty()) {
            auto node = stack.top();
            stack.pop();
            result.push_back(node->val);

            if (node->left != nullptr) stack.push(node->left);
            if (node->right != nullptr) stack.push(node->right);
        }

        std::reverse(result.begin(), result.end());
        return result;
    }

    std::vector<int> ans;
    void postorder(const TreeNode* root) {
        if(root==nullptr) return;

        postorder(root->left);
        postorder(root->right);
        ans.push_back(root->val);
    }
    std::vector<int> postorderTraversal(const TreeNode* root) {
        //left right root
        postorder(root);
        return ans;
    }
};

File: source/python/145. Binary Tree Postorder Traversal.py
from queue import LifoQueue
from typing import Self


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val: int = 0, left: Self | None = None, right: Self | None = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # def postorderTraversal(self, root: TreeNode | None) -> list[int]:
    #     if root is None:
    #         return []

    #     return self.postorderTraversal(root.left) + self.postorderTraversal(root.right) + [root.val]

    def postorderTraversal(self, root: TreeNode | None) -> list[int]:
        if root is None:
            return []

        stack: LifoQueue[TreeNode] = LifoQueue(maxsize=100)
        stack.put(root)

        result: list[int] = []

        while not stack.empty():
            node = stack.get()
            result.append(node.val)

            if node.left is not None:
                stack.put(node.left)
            if node.right is not None:
                stack.put(node.right)

        return result[::-1]


def test():
    s = Solution()
    assert s.postorderTraversal(
        TreeNode(1, None, TreeNode(2, TreeNode(3)))) == [3, 2, 1]
    assert s.postorderTraversal(None) == []
    assert s.postorderTraversal(TreeNode(1)) == [1]

    print("All tests passed!")


if __name__ == "__main__":
    test()
